### PR TITLE
Create einsum operation

### PR DESCRIPTION
**Pull request by MilesCranmer**
This creates the functional einsum function as requested on #73. CC @arogozhnikov @cgarciae 

The current implementation simply parses the string and converts it to `einsum` notation by mapping axis names to single characters (I use `string.ascii_letters`, starting from `a, b, c` etc).

Currently, it has the following features:

- Supports the backends: tensorflow, numpy, jax, pytorch, chainer, oneflow, keras, cupy.
- Allows for an arbitrary number of tensors passed.
- Allows ellipsis specification, including for multiple tensors, so long as it is provided on both the left and the right of the `->`.

It does not currently support
- Reshape operations, such as `"(batch channel) feature, feature -> batch channel"`.
- Custom reduction operations.

These could be added later if desired. Some backends do not support custom reductions in their einsum implementations so it will be a bit more work.

I also added a docstring and some unittests (in `tests/test_einsum.py`). 

Here are some examples of use, with the numpy backend:
```python
# Filter a set of images:
>>> batched_images = np.random.randn(128, 16, 16)
>>> filters = np.random.randn(16, 16, 30)
>>> result = einsum(batched_images, filters,
...                 "batch h w, h w channel -> batch channel") 

>>> result.shape
(128, 30)

# Matrix multiplication, with an unknown input shape:
>>> batch_shape = (50, 30)
>>> data = np.random.randn(*batch_shape, 20)
>>> weights = np.random.randn(10, 20)
>>> result = einsum(weights, data, 
...                 "out_dim in_dim, ... in_dim -> ... out_dim")
>>> result.shape
(50, 30, 10)
```

Note that the number of spaces next to the comma above are arbitrary, you could do either `"in_dim, ..."` or `"in_dim , ..."` - both will work.

Eager to hear feedback on this!

Cheers,
Miles

---

**Edit 1**: Got working for repeat indices on one side (as used in, e.g., trace).
**Edit 2**: Added support for chainer, oneflow, cupy, tensorflow.keras.
**Edit 3**: Added many more tests, some mirroring those used in the np.einsum tests.
**Edit 4**: More and more unit tests. 
**Edit 5**: Tweaked the syntax to have tensors first, pattern second. Adapted tests, and added new validation for order of arguments.

### CODE REVIEW

**Comment by MilesCranmer** [comment]
I think implementing the `rearrange` operations inside shouldn't be too bad: since einsum doesn't require any specific order of indices, you could call
```python
for (tensor, left_expression) in zip(tensors, left_expressions):
    axis_names = ...
    tensor = rearrange(left_expression + "->" + " ".join(axis_names))
```
 for every `left_expression`. Then, you would pass `" ".join(axis_names)` back to `einops.einsum`.

Then, you could do a similar rearrange on the output expression.

What do you think @arogozhnikov?

**Comment by MilesCranmer** [comment]
Actually I realized this doesn't work with repeat variable names. e.g., if I want to compute the trace of a tensor:
```python
einsum("index index -> ", np.ones((5, 5)))
```
`ParsedExpression` doesn't allow duplicate dimensions so this doesn't work. I guess I could modify it to allow this.

File: einops/einops.py [hunk 1]
```diff
@@ -624,3 +625,132 @@ def asnumpy(tensor) -> 'numpy.ndarray':
...
+
+    left_patterns = []
+    for left in lefts:
+        left_pattern = ""
+        for raw_axis_name in left.composition:
+
+            if raw_axis_name == _ellipsis:
+                left_pattern += '...'
+                continue
+            elif len(raw_axis_name) > 1:
```
Comment (arogozhnikov) [inline einops/einops.py:659]: != 1, so singleton ()-axes are reported here too.

File: einops/einops.py [hunk 2]
```diff
@@ -624,3 +625,132 @@ def asnumpy(tensor) -> 'numpy.ndarray':
         `numpy.ndarray`, converted to numpy
     """
     return get_backend(tensor).to_numpy(tensor)
+
+
+@functools.lru_cache(256)
+def _compactify_pattern_for_einsum(pattern: str) -> str:
+    if "->" not in pattern:
```
Comment (arogozhnikov) [inline einops/einops.py:632]: In several latest things I switched to using <- because output is most cases is on the left, and inputs are on the right. 

This deviates from original einops operations, but may be helpful here to make additional visible difference with native einsums.

Don't have any strong opinion on that, just curious to get your thoughts.

File: einops/einops.py [hunk 3]
```diff
@@ -624,3 +625,132 @@ def asnumpy(tensor) -> 'numpy.ndarray':
...
+        # numpy allows this, so make sure users
+        # don't accidentally do something like this.
+        raise ValueError("Einsum pattern must contain '->'.")
+    lefts, right = pattern.split('->')
+    lefts = lefts.split(',')
+
+    lefts = [
+        ParsedExpression(left, allow_underscore=True, allow_duplicates=True)
```
Comment (arogozhnikov) [inline einops/einops.py:640]: Looks like it will pass things like "b 1 h w->c 1" and assign "1" as one of axes.
Worth adding tests that parsing fails on the following cases:
"i 1 -> 1 i"
"i () -> i"
"i (j k) -> i j k"

File: einops/einops.py [hunk 4]
```diff
@@ -624,3 +625,132 @@ def asnumpy(tensor) -> 'numpy.ndarray':
...
+            left_pattern += axis_name_mapping[axis_name]
+        left_patterns.append(left_pattern)
+
+    output_pattern = ",".join(left_patterns) + "->"
+
+    for raw_axis_name in right.composition:
+        if raw_axis_name == _ellipsis:
+            output_pattern += '...'
+            continue
+        elif len(raw_axis_name) > 1:
```
Comment (arogozhnikov) [inline einops/einops.py:679]: again, likely != 1

File: einops/einops.py [hunk 5]
```diff
@@ -624,3 +625,132 @@ def asnumpy(tensor) -> 'numpy.ndarray':
...
+    ```
+
+    Parameters:
+        pattern: string, einsum pattern, with commas
+                 separating specifications for each tensor.
+        tensors: tensors of any supported library (numpy, tensorflow, pytorch, jax).
+
+    Returns:
+        Tensor of the same type as input, after processing with einsum.
+
```
Comment (arogozhnikov) [inline einops/einops.py:753]: nice detailed docstring

File: tests/test_einsum.py [hunk 6]
```diff
@@ -40,11 +41,131 @@ def __call__(self, function):
 ]
 
 
-def test_all():
+# Each of the form:
+# (Arguments, true_einsum_pattern, in_shapes, out_shape)
+test_functional_cases = [
```
Comment (arogozhnikov) [inline tests/test_einsum.py:46]: thank you for collecting various test cases

File: tests/test_einsum.py [hunk 7]
```diff
@@ -53,6 +174,47 @@ def test_all():
                 output = backend.to_numpy(output_framework)
                 assert output.shape == out_shape
 
+
+def test_functional():
+    # Functional tests:
+    valid_backends = ['tensorflow', 'torch', 'jax', 'numpy',
+                      'chainer', 'oneflow', 'cupy', 'tensorflow.keras']
+    for backend in collect_test_backends():
```
Comment (arogozhnikov) [inline tests/test_einsum.py:182]: important - this collects only non-symbolic frameworks.
Need one more similar section with symbolic frameworks, the only difference is computation part itself, see here:

https://github.com/arogozhnikov/einops/blob/master/tests/test_ops.py#L247-L250

**Review by arogozhnikov — COMMENTED** [review]
cool, thanks Miles for very clean PR and collecting test suite

I've left some thoughts and request for two kinds of tests: 
1) fail on parsing of features that aren't yet supported  
2) add tests for symbolic backends. Probably latter would only include tf, but still

File: einops/einops.py (hunk 2 above)
Comment (MilesCranmer) [reply einops/einops.py:632]: I'm fine with either. Although, if I am used to using `rearrange, repeat, reduce`, or even regular `einsum`, then the first syntax I would intuitively try would be `->`. I think you wish to encourage `<-` in the future, I would allow for both `<-` and `->` patterns in all functions for the moment, and maybe give a deprecation warning in a future version that users should switch to `<-` patterns.

Just glancing at `<-`, I would assume it is an inplace operation, whereas `->` creates a new tensor. It is a bit confusing though, I feel like we should just go with `->` for everything.

@shoyer thoughts?

File: tests/test_einsum.py (hunk 7 above)
Comment (MilesCranmer) [reply tests/test_einsum.py:184]: I just tried this, but it seems there aren't any available symbolic frameworks available for einsum. MXNet seems to not have einsum available for either their symbolic or non-symbolic backends. They have one listed under `mxnet.np.einsum`, but it uses a different internal backend incompatible with `mxnet.nd`. The other symbolic one is keras, but only with `layers=True`, which can't seem to be passed through the functional version of einsum.

File: einops/einops.py (hunk 3 above)
Comment (MilesCranmer) [reply einops/einops.py:640]: Good idea. Adding now.

File: tests/test_einsum.py (hunk 7 above)
Comment (MilesCranmer) [reply tests/test_einsum.py]: Actually it seems I might be able to get the Keras one working. But the symbols have one extra leading axis than expected. I guess for those, I need to vectorize the `tf.einsum` over its inputs by default?

Comment (MilesCranmer) [reply tests/test_einsum.py]: Okay got a symbolic test working for keras!

File: einops/einops.py (hunk 3 above)
Comment (MilesCranmer) [reply einops/einops.py:655]: Done.

File: einops/einops.py (hunk 2 above)
Comment (MilesCranmer) [reply einops/einops.py:647]: One other reason my eyes seem to prefer `->` is because the same symbol is used for type declarations in function output, and any einops index string is sort of like a type declaration: the left side represents the arguments, and the right side is the output type.

**Comment by MilesCranmer** [comment]
Okay, all suggestions implemented. Let me know what you think.

File: einops/einops.py (hunk 2 above)
Comment (MilesCranmer) [reply einops/einops.py:647]: I tweaked the syntax to be `y = einsum(x, x, "i j, i k -> j k")`. Let me know if this is okay.

**Comment by MilesCranmer** [comment]
Okay everything is implemented for the new syntax:
```python
y = einsum(x, x, "i j, i k -> j k")
```

I also added new validation checks for the argument order, and corresponding unit-tests. 

In the unittests, I also now check the specific message of each error, rather than the error type.

Let me know what you think.

File: einops/_backends.py [hunk 8]
```diff
@@ -524,9 +543,8 @@ def create_symbol(self, shape):
         return self.keras.Input(batch_shape=shape)
 
     def eval_symbol(self, symbol, input_dict):
-        (variable, value), = input_dict
-        model = self.keras.models.Model(variable, symbol)
-        return model.predict_on_batch(value)
+        model = self.keras.models.Model([var for (var, _) in input_dict], symbol)
```
[the remark this answers is not in the document]
Comment (arogozhnikov) [reply einops/_backends.py:546]: yes, that's completely fine

File: tests/test_einsum.py [hunk 9]
```diff
@@ -53,6 +177,181 @@ def test_all():
...
+                if predicted_out_data.shape != out_shape:
+                    raise ValueError(
+                        f"Expected output shape {out_shape} but got {predicted_out_data.shape}"
+                    )
+                assert np.testing.assert_array_almost_equal(predicted_out_data,
+                                                            expected_out_data,
+                                                            decimal=5)
+
+
+def test_functional_errors():
```
Comment (arogozhnikov) [inline tests/test_einsum.py:264]: very well documented 👍

**Comment by arogozhnikov** [comment]
> for every left_expression. Then, you would pass " ".join(axis_names) back to einops.einsum.
> Then, you could do a similar rearrange on the output expression.

It's trickier since you want some of axes to be derived from the inputs shapes.
In the examples you previously posted it could be like `(i j) k, i -> j k`, so first shape of the second argument should be parsed. For the output that's actually straighforward and just applying a pattern would always work

File: einops/_backends.py [hunk 10]
```diff
@@ -560,6 +578,12 @@ def layers(self):
         from .layers import keras
         return keras
 
+    def einsum(self, pattern, *x):
+        return self.tf.vectorized_map(
```
Comment (arogozhnikov) [inline einops/_backends.py:582]: want to understand why it looks so strange in tf.keras

Comment (MilesCranmer) [reply einops/_backends.py:582]: I'm not sure if I was interpreting the symbolic backends correctly or not. 

Basically, this `einsum` assumes the `x` tensors have a leading batch axis, which are assumed to not be specified in the pattern. I assumed this because the `create_symbol` method specifies the shape as a batch shape, rather than an absolute shape. Is that correct, or should it assume the pattern also specifies the batch axis?

File: tests/test_einsum.py (hunk 9 above)
Comment (MilesCranmer) [reply tests/test_einsum.py:264]: Thanks!

File: einops/_backends.py (hunk 10 above)
Comment (MilesCranmer) [reply einops/_backends.py:582]: Actually, I think my implementation has a potential issue: if one symbol is batched, and one symbol is not (like a weight matrix).

Comment (MilesCranmer) [reply einops/_backends.py:582]: What do you think the correct strategy is here? Should I avoid adding `einsum` for keras, since it is technically a `layer=True` backend?

Comment (arogozhnikov) [reply einops/_backends.py:582]: layer=True just refers to providing layers, it should not be related to any batch variables, and patterns should include batch variables. Anyway, I forgot keras now just a redirection to TF layers, so just excluded this part

**Comment by arogozhnikov** [comment]
PR is merged, made very minor changes.
thank you for paying attention to details and keeping pushing this!

**Comment by gerdm** [comment]
This is a great PR! 
+1 for “ Custom reduction operations”. Is there anyone already working on this?

**Comment by alok** [comment]
I'm interested in adding rearrange support. Any pointers?
